src/v/utils/retry_chain_node.h

```diff
@@ -147,8 +148,83 @@
 #include <fmt/ostream.h>
 
 #include <chrono>
+#include <exception>
+#include <iterator>
 #include <variant>
 
+/// The context of the retry chain node.
+/// This object can collect traces in memory.
+/// It uses limited amount of memory which can be configured.
+/// It can also provide an abort source that can be used to
+/// stop the fiber.
+template<typename Clock = ss::lowres_clock>
+class basic_retry_chain_context {
+public:
+    /// Create a trace memory buffer with the specified size.
+    explicit basic_retry_chain_context(ss::sstring name, ss::abort_source& as, size_t size)
+      : _name(std::move(name))
+      , _as(as)
+      , _size_limit(size) {}
+
+    /// Add a trace to the buffer.
+    void add_trace(const ss::sstring& msg) const {
+        _last_trace_ts = Clock::now();
+        if (_buf.size() >= _size_limit) {
+            if (!_is_truncated) {
+                // If the buffer is full, we truncate it and set the flag.
+                _is_truncated = true;
+                fmt::format_to(std::back_inserter(_buf), "[truncated]");
+            }
+            return;
+        }
+        auto bii = std::back_inserter(_buf);
+        std::copy(msg.begin(), msg.end(), bii);
+        bii = '\n';
+    }
```

Comment:
can you explain the semantics and motivation of those semantics of the logging in a class comment? for example, i thought we'd use a circular buffer so that when a stall occurred we'd see the last N messages. but this looks like once the buffer fills up we drop new messages. am i reading that correctly?

Reply:
I updated the comment.
The idea is that this mechanism should be used for periodic but limited logging. You can't log infinite amount of log messages between resets. If the other type of trace collection is needed another type of  context should be added.

Reply:
I think the value of this is very limited and the caller has to be very careful/conscious with buffer size/amount of logging. Maybe maintain a linked list of log lines with a total size? Would be pretty simple then to implement a fifo list of logs and it will be much easier to use/reason about.

Reply:
I replaced `fmt::memory_buffer` with the custom circular buffer implementation so now it is FIFO